### PR TITLE
FEAT(client): Message Thresholding

### PR DESCRIPTION
**Pull request by Skewjo**
### Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

This PR adds a new column named "Threshold" to the "Messages" settings tab to allow users to turn off message notifications when the server is above a certain number of users: 
image

### CODE REVIEW

File: src/mumble/Log.cpp [hunk 1]
```diff
@@ -717,14 +738,15 @@ void Log::log(MsgType mt, const QString &console, const QString &terse, bool own
 		}
 
 		// Message notification with static sounds
-		if ((flags & Settings::LogSoundfile)) {
+		if ((flags & Settings::LogSoundfile) && !(flags & Settings::LogMessageThreshold &&  g.sh.?? > 20 )) { //need current number of members in server here?
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:741]:
```suggestion
		int connectedUsers = 0;
		{
			QReadLocker lock(&ClientUser::c_qrwlUsers);
			connectedUsers = ClientUser::c_qmUsers.size();
		}
		if ((flags & Settings::LogSoundfile) && !(flags & Settings::LogMessageThreshold &&  connectedUsers > 20 )) {
```

I _think_ this should do the trick

Comment (Skewjo) [reply src/mumble/Log.cpp:741]: Ah! I almost got there, but I didn't use a lock. Thanks.

File: src/mumble/Log.cpp [hunk 2]
```diff
@@ -44,6 +44,7 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
 	qtwMessages->setAccessibleName(tr("Log messages"));
 	qsVolume->setAccessibleName(tr("TTS engine volume"));
 	qsbThreshold->setAccessibleName(tr("Length threshold"));
+	qsbMessageThresholdUsers->setAccessibleName(tr("Message threshold users"));
```
Comment (Skewjo) [inline src/mumble/Log.cpp:47]: Is putting this new variable here confusing in regards to the other unaffected variable right above it? (```qsbThreshold``` )

File: src/mumble/Log.ui [hunk 3]
```diff
@@ -193,7 +198,27 @@
...
+       </item>
+       <item>
+	    <widget class="QSpinBox" name="qsbMessageThresholdUsers">
+	     <property name="toolTip">
+		  <string>Number of users that will trigger thresholding functionality.</string>
+		  </property>
+		  <property name="frame">
+		   <bool>true</bool>
+		  </property>
+		 </widget>
```
Comment (Skewjo) [inline src/mumble/Log.ui:219]: This is spaces vs. tabs, huh? I need to change my visual studio settings to fix it I guess.

File: src/mumble/Log.ui [hunk 4]
```diff
@@ -193,7 +198,27 @@
...
+       </item>
+	 </layout>
```
Comment (Skewjo) [inline src/mumble/Log.ui:221]: Also, I tried to implement this like the following. 
```
	  <item>
        <widget class="QCheckBox" name="qcbMessageThreshold">
         <property name="toolTip">
          <string>If checked, notifications will not be played when more than 20 people are active on the server.</string>
         </property>
         <property name="text">
          <string>Don't play notifications when there are more than 
<widget class="QSpinBox" name="qsbMessageThresholdUsers">
	     <property name="toolTip">
		  <string>Number of users that will trigger thresholding functionality.</string>
		  </property>
		  <property name="frame">
		   <bool>true</bool>
		  </property>
		 </widget>
 number of users on the server.</string>
         </property>
		</widget>
       </item>
	 </layout>
```




It built/compiled fine, but crashed immediately when I opened the settings page. It obviously didn't like trying to put a spinbox widget in the middle of a string. I think I knew what to do next, but I didn't want to change the layout from "verticalLayout" to "gridLayout" (like the other settings box immediately below this one) prematurely.

**Comment by Skewjo** [comment]
Also, I haven't tested this functionality in any way, other than building & opening mumble to make sure the UI looked somewhat correct.

* Is there a way I can mock a bunch of users on a server?
    * If not, is there a well-known server I can jump into to check it?
* Do I need to write any unit tests for this functionality?

File: src/mumble/Log.ui (hunk 3 above)
Comment (Krzmbrzl) [reply src/mumble/Log.ui:219]: Are you editing these UI files manually or via Qt Designer?

File: src/mumble/Log.cpp (hunk 2 above)
Comment (Krzmbrzl) [reply src/mumble/Log.cpp:47]: Nah, I think it's fine

File: src/mumble/Log.ui (hunk 4 above)
Comment (Krzmbrzl) [reply src/mumble/Log.ui:221]: I can't help you with manually editing the UI files. Never did that and I strongly recommend against doing so if the changes are not trivial :point_up:

**Comment by Krzmbrzl** [comment]
> Is there a way I can mock a bunch of users on a server? 

You can use something like Botamusique to create a bunch of bots and let them join a locally hosted server

> If not, is there a well-known server I can jump into to check it?

If you set the threshold low enough and @Natenom allows it, you could use his server (I'll not give out any server details before he agreed though)

> Do I need to write any unit tests for this functionality?

As it stands we barely have any tests in the first place, so I don't see an easy way of adding test cases for this feature. We want to change that eventually but for now: no you don't have to

**Comment by Skewjo** [comment]
Also, here's what the new elements currently look like: 
image

File: src/mumble/Log.ui (hunk 3 above)
Comment (Skewjo) [reply src/mumble/Log.ui:219]: I did these manually...

:stuck_out_tongue_winking_eye:

**Comment by Krzmbrzl** [comment]
I think it'd look better if the threshold column was moved one position to the left so that the `Path` column remains the last column. That way all checkboxes would stay grouped together :point_up:

**Comment by Kissaki** [comment]
Can you please include a description of the (overall) changes and context in the PR description? This may be what is also written in an individual commit message. It should be enough context to get an overview for unfamiliar readers, and will land in the merge commit which again should be easy to digest to know what changed and why.

**Comment by Skewjo** [comment]
> Can you please include a description of the (overall) changes and context in the PR description? This may be what is also written in an individual commit message. It should be enough context to get an overview for unfamiliar readers, and will land in the merge commit which again should be easy to digest to know what changed and why.

Yes, I'm going to clean up the commits after I'm done with my changes. It's just auto-pushing from my fork :/

**Comment by Skewjo** [comment]
@Krzmbrzl 
I, mistakenly, installed a visual studio update before finishing this PR, and now my build process stays stuck at ```-- Qt5 component found: Widgets | Version: 5.15.2``` for an hour or more (I've attempted to rebuild it 3 times now). ¯\_(ツ)_/¯
So close, yet so far.
image

**Comment by davidebeatrici** [comment]
Prune the build directory.

**Comment by Krzmbrzl** [comment]
And also try waiting. Start the build and then go have lunch or something. On Windows it can take forever for this to work out.

**Comment by Skewjo** [comment]
@davidebeatrici, sorry, not sure how to prune the build directory. Is pruning referring to modifying one of the ```CMakeLists.txt``` files or a special ```prune``` command I can pass to CMake maybe?

**Comment by Skewjo** [comment]
This is about to make me lose my mind.

I can't figure out why the threshold values won't save, and I can't get the project to build in QT or Visual Studio so that I can properly attach a debugger. 
image

There are 2 (identical?) code blocks in the save and load functions that I thought would take care of it without requiring any modification from me, but that doesn't appear to be the case.
```
	settings_ptr->beginWriteArray(QLatin1String("messages"));
	for (QMap< int, quint32 >::const_iterator it = qmMessages.constBegin(); it != qmMessages.constEnd(); ++it) {
		settings_ptr->setArrayIndex(it.key());
		SAVELOAD(qmMessages[it.key()], "log");
	}
```
Any help would be greatly appreciated.

**Comment by Skewjo** [comment]
It's little things like this that are about to make me give up. 
For some reason Visual Studio has now decided that a ton of object types in Log.cpp are now undefined... but it has no issues taking me to the header file where it's defined when I right-click and "peek-definition".
image
This wasn't an issue last week, but apparently something has happened to break it.

**Comment by Skewjo** [comment]
Ok... I've built it, the threshold settings save, and it works (but not for server connect and disconnect).

**Comment by Kissaki** [comment]
Try removing the out and src folder, and check them out again. This should remove any intermediate and broken files.

Then generate and make steps again.

**Comment by Skewjo** [comment]
I'd like to see this polished a bit, but if I don't submit this PR now, I'm never going to finish it. I've checked the functionality for 2 different message types, but not all of them.

* This functionality does not work for "Server Connected" or "Server Disconnected". (I'm guessing the functionality I'm using to get the current number of users on the server can't run before you're connected (or after you're disconnected). Those should likely be greyed out. 
* I'd like to make the qCheckBox (```qcbMessageThreshold```) and qSpinBox (```qsbMessageThresholdUsers```) part of the same UI component, but I believe that would involve changing the layout of the Misc. box from ```layout class="QVBoxLayout""``` to ```layout class="QGridLayout"``` but that felt like an overreach.
    * The spinbox should probably be greyed out when the checkbox is unchecked.
* I'm worried the name of the column ("threshold") is too generic and may get confused with other functionalities.

**Comment by Krzmbrzl** [comment]
> This functionality does not work for "Server Connected" or "Server Disconnected". (I'm guessing the functionality I'm using to get the current number of users on the server can't run before you're connected (or after you're disconnected). Those should likely be greyed out.

It's definitely not possible for the `Connected` situation. And for the `Disconnected` one I can imagine the disconnect already having happened when this part of the code is reached.
But then again you don't get spammed by these messages anyways, since you will only get them once each.

> I'd like to make the qCheckBox (qcbMessageThreshold) and qSpinBox (qsbMessageThresholdUsers) part of the same UI component, but I believe that would involve changing the layout of the Misc. box from layout class="QVBoxLayout"" to layout class="QGridLayout" but that felt like an overreach. 

I think that'd be quite important to do in order to clarify the context. So changing the layout is fine

> The spinbox should probably be greyed out when the checkbox is unchecked.

Yes

> I'm worried the name of the column ("threshold") is too generic and may get confused with other functionalities.

It is rather cryptic, yes. I am having a hard time finding a better column header though. Maybe `Limit` might be better (with a tooltip of what exactly is being limited)?

**Comment by Krzmbrzl** [comment]
Note also that you have merge conflicts.

In order to resolve them, run `git pull --rebase https://github.com/mumble-voip/mumble.git master` and then resolve the merge conflicts as to https://www.freecodecamp.org/news/how-to-handle-merge-conflicts-in-git/

**Comment by Skewjo** [comment]
@Krzmbrzl I'm currently unable to figure out how to create new signal handlers in order to make the button gray-out. 
I found a signal handling function in ```Log.cpp``` to use as an example (```on_qtwMessages_itemChanged```), but it's only used in the following file: ```mumble\build\release\src\mumble\mumble_autogen\EWIEGA46WW\moc_Log.cpp```
...which as you can see is an autogen file, so I don't think I'm supposed to manually add it there. I also can't find any place in the Qt designer UI to add a "message handler" (signal handler?).

In Visual Studio I could add it from the UI editor or add it manually via the message map, but I can't find any similar options here.

**Comment by Skewjo** [comment]
@Krzmbrzl 

Holy guacamole, I think I'm finally done. 
1. Change to grid layout so the 2 buttons and text can be aligned :white_check_mark:
2. Make spinbox gray out when option is disabled :white_check_mark:
3. Added additional clarifying sentence to "Threshold" tooltip :white_check_mark:

File: src/mumble/Log.cpp [hunk 5]
```diff
@@ -64,8 +67,12 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
 	allMessagesItem->setToolTip(ColNotification, QObject::tr("Toggle pop-up notifications for all events"));
 	allMessagesItem->setCheckState(ColHighlight, Qt::Unchecked);
 	allMessagesItem->setToolTip(ColHighlight, QObject::tr("Toggle window highlight (if not active) for all events"));
+	allMessagesItem->setCheckState(ColMessageThreshold, Qt::Unchecked);
+    allMessagesItem->setToolTip(ColMessageThreshold, QObject::tr("Click here to toggle message thresholding for all events - Only works if the notification limiting option below is enabled."));
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:71]:
```suggestion
	allMessagesItem->setToolTip(ColMessageThreshold, QObject::tr("Enable limiting for all message types. Only effective if limiting is turned on in general."));
```

File: src/mumble/Log.cpp [hunk 6]
```diff
@@ -82,15 +89,19 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
...
 		twi->setCheckState(ColNotification, Qt::Unchecked);
 		twi->setCheckState(ColHighlight, Qt::Unchecked);
+		twi->setCheckState(ColMessageThreshold, Qt::Unchecked);
 		twi->setCheckState(ColStaticSound, Qt::Unchecked);
+		
 
 		twi->setToolTip(ColConsole, tr("Toggle console for %1 events").arg(messageName));
 		twi->setToolTip(ColNotification, tr("Toggle pop-up notifications for %1 events").arg(messageName));
 		twi->setToolTip(ColHighlight, tr("Toggle window highlight (if not active) for %1 events").arg(messageName));
+		twi->setToolTip(ColMessageThreshold, tr("Toggle thresholding behavior for %1 events ").arg(messageName));
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:99]:
```suggestion
		twi->setToolTip(ColMessageThreshold, tr("Toggle limiting for %1 events ").arg(messageName));
```

File: src/mumble/Log.cpp [hunk 7]
```diff
@@ -102,15 +113,21 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
 		twi->setWhatsThis(ColHighlight, tr("Click here to toggle window highlight for %1 events.<br />If checked, "
 										   "Mumble's window will be highlighted for every %1 event, if not active.")
 											.arg(messageName));
+		twi->setWhatsThis(ColMessageThreshold,
+			tr("Click here to toggle thresholding for %1 events. <br /> If checked, notifications for this event type"
+			   "will not be played when above the set number of users.<br />Ensure that the thresholding "
+			   "option below is enabled or this field will not have any effect.")
+				.arg(messageName));
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:120]:
```suggestion
		twi->setWhatsThis(ColMessageThreshold,
			tr("Click here to toggle limiting for %1 events. <br /> If checked, notification sounds for this event type"
			   "will not be played when the server exceeds the set number of connected users.<br />Ensure that the limit "
			   "option below is enabled or this field will not have any effect.")
				.arg(messageName));
```

File: src/mumble/Log.cpp [hunk 8]
```diff
@@ -102,15 +113,21 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
...
 		twi->setWhatsThis(ColStaticSound, tr("Click here to toggle sound notification for %1 events.<br />If checked, "
 											 "Mumble uses a sound file predefined by you to indicate %1 events. Sound "
 											 "files and Text-To-Speech cannot be used at the same time.")
 											  .arg(messageName));
 		twi->setWhatsThis(ColStaticSoundPath,
-						  tr("Path to sound file used for sound notifications in the case of %1 events.<br />Single "
+						 tr("Path to sound file used for sound notifications in the case of %1 events.<br />Single "
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:126]: Unrelated change

File: src/mumble/Log.cpp [hunk 9]
```diff
@@ -204,10 +226,11 @@ void LogConfig::load(const Settings &r) {
 #ifndef USE_NO_TTS
 		i->setCheckState(ColTTS, (ml & Settings::LogTTS) ? Qt::Checked : Qt::Unchecked);
 #endif
+		i->setCheckState(ColMessageThreshold, (ml & Settings::LogMessageThreshold) ? Qt::Checked : Qt::Unchecked);
 		i->setCheckState(ColStaticSound, (ml & Settings::LogSoundfile) ? Qt::Checked : Qt::Unchecked);
 		i->setText(ColStaticSoundPath, r.qmMessageSounds.value(mt));
 	}
-
+
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:233]: Unrelated change

File: src/mumble/Log.ui [hunk 10]
```diff
@@ -178,12 +183,22 @@
...
-      <string>Whisper</string>
+      <string>Misc.</string>
      </property>
-     <layout class="QVBoxLayout" name="verticalLayout">
-      <item>
+     <layout class="QGridLayout" name="gridLayout">
+      <item row="1" column="0">
+       <widget class="QCheckBox" name="qcbMessageThreshold">
+        <property name="toolTip">
+         <string>If checked, notifications will not be played when more than the defined number of users are active on the server.</string>
```
Comment (Krzmbrzl) [inline src/mumble/Log.ui:194]:
```suggestion
         <string>If checked, notification sounds will not be played when more than the defined number of users are connected to the server.</string>
```

File: src/mumble/Log.ui [hunk 11]
```diff
@@ -193,6 +208,48 @@
         </property>
        </widget>
       </item>
+      <item row="1" column="1">
+       <widget class="QSpinBox" name="qsbMessageThresholdUsers">
+        <property name="enabled">
```
Comment (Krzmbrzl) [inline src/mumble/Log.ui:213]: This has to be initialized depending on whether the corresponding checkbox is checked or not. Currently it always starts as enabled even if the checkbox is not checked. Only after checking and un-checking the checkbox, will this get disabled.

File: src/mumble/Log.ui [hunk 12]
```diff
@@ -178,12 +183,22 @@
...
-     <layout class="QVBoxLayout" name="verticalLayout">
-      <item>
+     <layout class="QGridLayout" name="gridLayout">
+      <item row="1" column="0">
+       <widget class="QCheckBox" name="qcbMessageThreshold">
+        <property name="toolTip">
+         <string>If checked, notifications will not be played when more than the defined number of users are active on the server.</string>
+        </property>
+        <property name="text">
+         <string>Don't play notifications when there are more than</string>
```
Comment (Krzmbrzl) [inline src/mumble/Log.ui:197]:
```suggestion
         <string>Limit notification sounds when more than</string>
```

File: src/mumble/Log.ui [hunk 13]
```diff
@@ -193,6 +208,48 @@
...
+        </property>
+        <property name="frame">
+         <bool>true</bool>
+        </property>
+       </widget>
+      </item>
+      <item row="1" column="2">
+       <widget class="QLabel" name="label">
+        <property name="text">
+         <string>users on the server.</string>
```
Comment (Krzmbrzl) [inline src/mumble/Log.ui:233]:
```suggestion
         <string>users are on the server.</string>
```

**Review by Krzmbrzl — CHANGES_REQUESTED** [review]
I think "Limit" is probably the better name for this feature.
a) It seems to be more easily understandable
b) It's shorter and thus the column header doesn't require as much space

In that spirit, I think you should rename all variables, setting keys and all UI strings that I might have overlooked to use "limit" instead of "threshold". Unless you disagree with that choice of course :point_up:

File: src/mumble/Log.cpp (hunk 8 above)
Comment (Skewjo) [reply src/mumble/Log.cpp:126]: You guys don't follow the "boy scout rule"? :stuck_out_tongue:

File: src/mumble/Log.ui (hunk 11 above)
Comment (Skewjo) [reply src/mumble/Log.ui:213]: I could've sword I check that, and I thought the signal/slot functionality was handling it for me...

File: src/mumble/Log.cpp (hunk 8 above)
Comment (Krzmbrzl) [reply src/mumble/Log.cpp:126]: It depends. Fixing this is fine, but it should be separated from the main changes. So if you wrap this up in its own `FORMAT` commit, then that'd be fine :)

**Comment by Skewjo** [comment]
Oh sorry, I thought I already changed that column header to be "Limit". Let me look again.

**Comment by Skewjo** [comment]
Yep, just needed to change the screenshot.

**Comment by Krzmbrzl** [comment]
> There also seems to be no connection between the column and the setting below. I’m not sure if they are supposed to be connected or alternatives, but either way if they represent the same functionality they should use the same terminology.

Yep that is a good point. We should probably simply have the checkboxes in the column and below only have the spin box for setting the user count at which limiting starts :thinking:

**Comment by Skewjo** [comment]
> > There also seems to be no connection between the column and the setting below. I’m not sure if they are supposed to be connected or alternatives, but either way if they represent the same functionality they should use the same terminology.
> 
> Yep that is a good point. We should probably simply have the checkboxes in the column and below only have the spin box for setting the user count at which limiting starts 🤔

That should be easy enough.

**Comment by Skewjo** [comment]
Ok @Krzmbrzl, mind giving it another look?

File: src/mumble/Log.cpp [hunk 14]
```diff
@@ -41,6 +41,7 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
 	qtwMessages->setAccessibleName(tr("Log messages"));
 	qsVolume->setAccessibleName(tr("TTS engine volume"));
 	qsbThreshold->setAccessibleName(tr("Length threshold"));
+	qsbMessageLimitUsers->setAccessibleName(tr("Message limit users"));
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:44]:
```suggestion
	qsbMessageLimitUsers->setAccessibleName(tr("User limit for message limiting"));
```

File: src/mumble/Log.cpp [hunk 15]
```diff
@@ -64,6 +66,8 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
 	allMessagesItem->setToolTip(ColNotification, QObject::tr("Toggle pop-up notifications for all events"));
 	allMessagesItem->setCheckState(ColHighlight, Qt::Unchecked);
 	allMessagesItem->setToolTip(ColHighlight, QObject::tr("Toggle window highlight (if not active) for all events"));
+	allMessagesItem->setCheckState(ColMessageLimit, Qt::Unchecked);
+	allMessagesItem->setToolTip(ColMessageLimit, QObject::tr("Click here to toggle message limiting for all events - If using this option be sure to change the user limit below."));
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:70]:
```suggestion
	allMessagesItem->setToolTip(ColMessageLimit, tr("Click here to toggle message limiting for all events - If using this option be sure to change the user limit below."));
```
We recently figure out that one should not use `QObject::` to access this function as Qt performas some behind-the-scene magic that involves the scope at which this function is invoked.

File: src/mumble/Log.cpp [hunk 16]
```diff
@@ -102,6 +108,11 @@ LogConfig::LogConfig(Settings &st) : ConfigWidget(st) {
 		twi->setWhatsThis(ColHighlight, tr("Click here to toggle window highlight for %1 events.<br />If checked, "
 										   "Mumble's window will be highlighted for every %1 event, if not active.")
 											.arg(messageName));
+		twi->setWhatsThis(ColMessageLimit,
+						  tr("Click here to toggle limiting for %1 events. <br /> If checked, notifications for this event type"
+							 "will not be played when above the set number of users.<br />Ensure that the limiting "
+							 "option below is enabled or this field will not have any effect.")
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:114]:
```suggestion
							 "will not be played when the user count on the server exceeds the set threshold."
							)
```

File: src/mumble/Log.cpp [hunk 17]
```diff
@@ -152,6 +164,9 @@ void LogConfig::updateSelectAllButtons() {
 			allTTSChecked = false;
 		}
 #endif
+	if (i->checkState(ColMessageLimit) != Qt::Checked) {
+	    allLimitChecked = false;
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:168]:
```suggestion
		if (i->checkState(ColMessageLimit) != Qt::Checked) {
			allLimitChecked = false;
```

File: src/mumble/Log.cpp [hunk 18]
```diff
@@ -204,6 +220,7 @@ void LogConfig::load(const Settings &r) {
 #ifndef USE_NO_TTS
 		i->setCheckState(ColTTS, (ml & Settings::LogTTS) ? Qt::Checked : Qt::Unchecked);
 #endif
+	i->setCheckState(ColMessageLimit, (ml & Settings::LogMessageLimit) ? Qt::Checked : Qt::Unchecked);
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:223]:
```suggestion
		i->setCheckState(ColMessageLimit, (ml & Settings::LogMessageLimit) ? Qt::Checked : Qt::Unchecked);
```

File: src/mumble/Log.cpp [hunk 19]
```diff
@@ -244,6 +262,8 @@ void LogConfig::save() const {
 		if (i->checkState(ColTTS) == Qt::Checked)
 			v |= Settings::LogTTS;
 #endif
+	if (i->checkState(ColMessageLimit) == Qt::Checked)
+	    v |= Settings::LogMessageLimit;
```
Comment (Krzmbrzl) [inline src/mumble/Log.cpp:266]:
```suggestion
			if (i->checkState(ColMessageLimit) == Qt::Checked) {
				v |= Settings::LogMessageLimit;
			}
```

File: src/mumble/Settings.cpp [hunk 20]
```diff
@@ -512,7 +513,7 @@ Settings::Settings() {
 	bEnableUIAccess             = true;
 
 	for (int i = Log::firstMsgType; i <= Log::lastMsgType; ++i) {
-		qmMessages.insert(i, Settings::LogConsole | Settings::LogBalloon | Settings::LogTTS);
+	qmMessages.insert(i, Settings::LogConsole | Settings::LogBalloon | Settings::LogTTS | Settings::LogMessageLimit);
```
Comment (Krzmbrzl) [inline src/mumble/Settings.cpp:516]:
```suggestion
		qmMessages.insert(i, Settings::LogConsole | Settings::LogBalloon | Settings::LogTTS | Settings::LogMessageLimit);
```
Doesn't this by default enable message limiting for all message types? I think in this case we should default to this feature being disabled.

File: src/mumble/Settings.h [hunk 21]
```diff
@@ -170,6 +170,7 @@ struct Settings {
 	bool bTTS;
 	bool bUserTop;
 	bool bWhisperFriends;
+	int iMessageLimitUsers;
```
Comment (Krzmbrzl) [inline src/mumble/Settings.h:173]:
```suggestion
	int iMessageLimitUserThreshold;
```

File: src/mumble/Settings.cpp [hunk 22]
```diff
@@ -335,6 +335,7 @@ Settings::Settings() {
 	bUserTop = true;
 
 	bWhisperFriends = false;
+	iMessageLimitUsers = 0;
```
Comment (Krzmbrzl) [inline src/mumble/Settings.cpp:338]:
```suggestion
	iMessageLimitUsers = 20;
```
This seems like a more sensible default than `0`.

File: src/mumble/Settings.h [hunk 23]
```diff
@@ -281,8 +282,10 @@ struct Settings {
 		LogTTS       = 0x02,
 		LogBalloon   = 0x04,
 		LogSoundfile = 0x08,
-		LogHighlight = 0x10
+		LogHighlight = 0x10,
+		LogMessageLimit = 0x20
```
Comment (Krzmbrzl) [inline src/mumble/Settings.h:286]:
```suggestion
		LogMessageLimit = 0x20,
```

**Review by Krzmbrzl — CHANGES_REQUESTED** [review]
Furthermore I think the menu entry for setting the user threshold for the limiting feature should be a simple form. Something like
```
Notification limit threshold: <spinBox> users
```

Also sorry for the delay. I didn't get to review this earlier.

**Comment by Skewjo** [comment]
Ok, @Krzmbrzl, I think we're good now. Sorry for the delay. Got a new job and have limited time to work on this issue.

**Comment by Skewjo** [comment]
Ok @Krzmbrzl, I've got 2 days to finish this thing. Any chance you could take a look at it some time today?

**Review by Krzmbrzl — APPROVED** [review]
The changes LGTM.

The only thing I notices was that when you select the Limit option for all messages, all checkboxes are checked as epxteced. However unchecking a single checkbox for a particular message then causes all boxes to go unnchecked again.
This is a general problem of that settings page though and as such unrelated to your changes. Thus that is fine :)

I rebased your PR against the latest master branch and also added the missing translation commit.

I assume you have tested these changes to work as expected?

**Comment by Krzmbrzl** [comment]
Thank you very much for your contribution :+1:

**Comment by Skewjo** [comment]
You're welcome! Glad to have been able to help. Hoping to make some time to do some documentation PRs some time in the near future.
